pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -158,23 +199,50 @@ func ingessNodeFwAttach(objs bpfObjects, ifacesName []string, pinDir string, isD
 				log.Fatalf("could not attach XDP program: %s", err)
 				return err
 			}
-
-			lpinDir := path.Join(pinDir, ifaceName+"-link")
-			if err := l.Pin(lpinDir); err != nil {
-				log.Fatalf("failed to pin link to pinDir %s: %s", lpinDir, err)
+			lPinDir := path.Join(infc.pinPath, ifaceName+"_link")
+			if err := l.Pin(lPinDir); err != nil {
+				log.Fatalf("failed to pin link to pinDir %s: %s", lPinDir, err)
 				return err
 			}
+			infc.links = append(infc.links, l)
 			log.Printf("Attached IngressNode Firewall program to iface %q (index %d)", iface.Name, iface.Index)
 		} else {
 			log.Printf("Unattaching IngressNode Firewall program from iface %q (index %d)", iface.Name, iface.Index)
-			l, _ := link.AttachXDP(link.XDPOptions{
-				Program:   objs.IngresNodeFirewallProcess,
-				Interface: iface.Index,
-			})
-			l.Unpin()
-			l.Close()
-			objs.Close()
+			infc.cleanup()
 		}
 	}
 	return nil
 }
+
+// cleanup will delete all link objects for all interface and remove all the maps
+func (infc *IngNodeFwController) cleanup() {
+	for _, l := range infc.links {
+		l := l
+		l.Unpin()
+		l.Close()
+	}
+	infc.objs.Close()
+}
+
+func parseDstPorts(ports string) (uint16, uint16, error) {
+	if !strings.Contains(ports, "-") {
+		port, err := strconv.ParseUint(ports, 10, 16)
+		if err != nil {
+			return 0, 0, fmt.Errorf("invalid DstPort number %s", err)
+		}
+		return uint16(port), 0, nil
+	}
+	ps := strings.SplitN(ports, "-", 2)
+	if len(ps) != 2 {
+		return 0, 0, fmt.Errorf("invalid DstPorts range %q", ports)
+	}
+	startPort, err := strconv.ParseUint(ps[0], 10, 16)
+	if err != nil {
+		return 0, 0, fmt.Errorf("invalid Start DstPort number %s", err)
+	}
+	endPort, err := strconv.ParseUint(ps[1], 10, 16)
+	if err != nil {
+		return 0, 0, fmt.Errorf("invalid End DstPort number %s", err)
+	}
+	return uint16(startPort), uint16(endPort), nil
```

Comment:
Probably should validate ports are within range before casting? Defence in depth.

Reply:
wouldn't ParseUint with bitSize `16` will fail if the value larger than what u16 can hold ?